src/gui/src/PatternEditor/PianoRollEditor.cpp

```diff
@@ -729,32 +856,132 @@ void PianoRollEditor::addOrDeleteNoteAction( int nColumn,
 			pPattern->insert_note( pNote );
-		}
-
-		// hear note
-		Preferences *pref = Preferences::get_instance();
-		if ( pref->getHearNewNotes() && !noteOff ) {
-			Note *pNote2 = new Note( pSelectedInstrument, 0, fVelocity, fPan_L, fPan_R, nLength, fPitch);
-			pNote2->set_key_octave( pressednotekey, pressedoctave );
-			AudioEngine::get_instance()->get_sampler()->note_on(pNote2);
+			if ( m_bSelectNewNotes ) {
+				m_selection.addToSelection( pNote );
+			}
 		}
 	}
 	pSong->set_is_modified( true );
 	AudioEngine::get_instance()->unlock(); // unlock the audio engine
 
-	updateEditor();
+	updateEditor( true );
 	m_pPatternEditorPanel->getVelocityEditor()->updateEditor();
 	m_pPatternEditorPanel->getPanEditor()->updateEditor();
 	m_pPatternEditorPanel->getLeadLagEditor()->updateEditor();
 	m_pPatternEditorPanel->getNoteKeyEditor()->updateEditor();
 	m_pPatternEditorPanel->getDrumPatternEditor()->updateEditor();
 }
 
+void PianoRollEditor::updateModifiers( QInputEvent *ev ) {
+	// Key: Alt + drag: move notes with fine-grained positioning
+	m_bFineGrained = ev->modifiers() & Qt::AltModifier;
+	// Key: Ctrl + drag: copy notes rather than moving
+	m_bCopyNotMove = ev->modifiers() & Qt::ControlModifier;
+
+	if ( m_selection.isMoving() ) {
+		// If a selection is currently being moved, change the cursor
+		// appropriately. Selection will change it back after the move
+		// is complete (or abandoned)
+		if ( m_bCopyNotMove &&  cursor().shape() != Qt::DragCopyCursor ) {
+			setCursor( QCursor( Qt::DragCopyCursor ) );
+		} else if ( !m_bCopyNotMove && cursor().shape() != Qt::DragMoveCursor ) {
+			setCursor( QCursor( Qt::DragMoveCursor ) );
+		}
+	}
+}
+
+
+QPoint PianoRollEditor::movingGridOffset( ) {
+	QPoint rawOffset = m_selection.movingOffset();
+	// Quantize offset to multiples of m_nGrid{Width,Height}
+	int nQuantX = m_nGridWidth, nQuantY = m_nRowHeight;
+	float nFactor = 1;
+	if ( ! m_bFineGrained ) {
+		int nBase = m_bUseTriplets ? 3 : 4;
+		nFactor = (4 * MAX_NOTES) / (nBase * m_nResolution);
+		nQuantX = m_nGridWidth * nFactor;
+	}
+	int x_bias = nQuantX / 2, y_bias = nQuantY / 2;
+	if ( rawOffset.y() < 0 ) {
+		y_bias = -y_bias;
+	}
+	if ( rawOffset.x() < 0 ) {
+		x_bias = -x_bias;
+	}
+	int x_off = (rawOffset.x() + x_bias) / nQuantX;
+	int y_off = (rawOffset.y() + y_bias) / nQuantY;
+	return QPoint( nFactor * x_off, y_off);
+}
+
+// Find a note that matches pNote, and move it from (nColumn, nRow) to (nNewColumn, nNewRow)
+void PianoRollEditor::moveNoteAction( int nColumn,
+									  Note::Octave octave,
+									  Note::Key key,
+									  int nPattern,
+									  int nNewColumn,
+									  Note::Octave newOctave,
+									  Note::Key newKey,
+									  Note *pNote)
+{
+	Hydrogen *pHydrogen = Hydrogen::get_instance();
+	Song *pSong = pHydrogen->getSong();
+
+	AudioEngine::get_instance()->lock( RIGHT_HERE );
+	PatternList *pPatternList = pSong->get_pattern_list();
+	InstrumentList *pInstrumentList = pSong->get_instrument_list();
```

Comment:
clang says this value is never read, which seems to the case :)

Reply:
>     * After selecting notes in the pattern editor, i could find no easy way to "disband" the selection again. Most other programs would do that after the clicks in another area (outside of the selection), but this just placed new notes in the pattern editor. Is it possible to do this without creating a new note?

Hmm, yes, this sounds like it's the right thing to do. Excellent! Will do, thanks.

Thanks for the review comments!